src/fu-device.c

```diff
@@ -164,6 +165,50 @@ fu_device_set_alternate (FuDevice *device, FuDevice *alternate)
 	g_set_object (&priv->alternate, alternate);
 }
 
+/**
+ * fu_device_get_parent:
+ * @device: A #FuDevice
+ *
+ * Gets any parent device. An parent device is logically "above" the current
+ * device and this may be reflected in client tools.
+ *
+ * This information also allows the plugin to optionally verify the parent
+ * device, for instance checking the parent device firmware version.
+ *
+ * Returns: (transfer none): a #FuDevice or %NULL
+ *
+ * Since: 1.0.8
+ **/
+FuDevice *
+fu_device_get_parent (FuDevice *device)
+{
+	FuDevicePrivate *priv = GET_PRIVATE (device);
+	g_return_val_if_fail (FU_IS_DEVICE (device), NULL);
+	return priv->parent;
+}
+
+/**
+ * fu_device_set_parent:
+ * @device: A #FuDevice
+ * @parent: Another #FuDevice
+ *
+ * Sets any parent device. An parent device is logically linked to the primary
+ * device in some way.
+ *
+ * Since: 1.0.8
+ **/
+void
+fu_device_set_parent (FuDevice *device, FuDevice *parent)
+{
```

Comment:
I think I would rather see these send and return parent ID's and let the daemon look up corresponding FuDevice. My logic is that it will be easier to predict and associate cross plugin which device is your parent if you know how the device is logically built or connected in advance.

Reply:
Yes and no. I do think having the FuDevice `GObject` of the parent is a useful thing to plugins, but I agree the daemon could do more matching up devices. I was intending to use `fu_device_set_parent()` directly in plugins where all the logical devices are in the "same" plugin (e.g. `unifying`) and for where the logical devices are spread around between plugins I'd add something like `fu_device_set_parent_with_guid(d,guid_string)` and/or `fu_device_set_parent_with_metdata(d,key,value)` and /then/ let the daemon call `fu_device_set_parent` so the actual object is usable by the "child" plugin when and if the parent shows up.

Reply:
Okay that works fine to me then.

Reply:
I was thinking more about the "find me a parent" case. I guess the most useful thing would be `fu_device_set_parent_with_guid` where the GUID is stored in the child `FuDevice` in an array (in the case the parent might be different things). Then we have two cases:

* parent added, then child
* child added, then parent

In both cases we just need to rescan the "adoption" list when a device is added. I think it's fair to say we won't allow re-parenting or having more than one parent. Would this work for your dock case?

Reply:
Yeah I think that would work.

Reply:
Something like the patch below?